File: src/main/java/com/fish/extendedae_plus_client/integration/recipeViewer/HelperRecipeViewer.java
package com.fish.extendedae_plus_client.integration.recipeViewer;

import appeng.api.stacks.GenericStack;
import com.fish.extendedae_plus_client.integration.ContextModLoaded;
import com.fish.extendedae_plus_client.integration.recipeViewer.emi.ViewerEmi;
import com.fish.extendedae_plus_client.integration.recipeViewer.jei.ViewerJei;
import com.mojang.datafixers.util.Pair;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Optional;

public class HelperRecipeViewer {
    private static IRecipeViewer activeViewer;

    public static void init() {
        if (ContextModLoaded.emi.isLoaded()) activeViewer = new ViewerEmi();
        else if (ContextModLoaded.jei.isLoaded()) activeViewer = new ViewerJei();
        else activeViewer = new EmptyHelper();
    }

    public static Optional<IRecipeViewer> getViewer() {
        if (activeViewer == null) init();
        return Optional.ofNullable(activeViewer);
    }

    public static List<GenericStack> getHoveredStacks() {
        return getViewer().map(IRecipeViewer::getHoveredStacks).orElse(List.of());
    }

    public static List<GenericStack> getFavorites() {
        return getViewer().map(IRecipeViewer::getFavorites).orElse(List.of());
    }

    /// @return Pair<Boolean: Stack, Boolean: ToInv>
    public static @Nullable Pair<Boolean, Boolean> matchesKey(int mouseKey) {
        return getViewer().map(viewer -> viewer.matchesKey(mouseKey)).orElse(null);
    }

    public static boolean isCheatMode() {
        return getViewer().map(IRecipeViewer::isCheatMode).orElse(false);
    }

    public static void addFavorite(GenericStack stack) {
        getViewer().ifPresent(viewer -> viewer.addFavorite(stack));
    }

    public static void setSearchText(String text) {
        getViewer().ifPresent(viewer -> viewer.setSearch(text));
    }
}

File: src/main/kotlin/com/fish/extendedae_plus_client/integration/recipeViewer/HelperRecipeViewer.kt
package com.fish.extendedae_plus_client.integration.recipeViewer

import appeng.api.stacks.GenericStack
import com.fish.extendedae_plus_client.integration.ContextModLoaded
import com.fish.extendedae_plus_client.integration.recipeViewer.emi.ViewerEmi
import com.fish.extendedae_plus_client.integration.recipeViewer.jei.ViewerJei
import com.mojang.datafixers.util.Pair
import java.util.*

object HelperRecipeViewer {
    private var activeViewer: IRecipeViewer? = null

    fun init() {
        activeViewer = if (ContextModLoaded.emi.isLoaded) ViewerEmi()
        else if (ContextModLoaded.jei.isLoaded) ViewerJei()
        else ViewerEmpty()
    }

    val viewer: Optional<IRecipeViewer>
        get() {
            if (activeViewer == null) init()
            return Optional.ofNullable<IRecipeViewer>(activeViewer)
        }

    @JvmStatic
    val hoveredStacks: MutableList<GenericStack?>
        get() = viewer
            .map(IRecipeViewer::hoveredStacks)
            .orElse(mutableListOf())

    @JvmStatic
    val favorites: MutableList<GenericStack?>
        get() = viewer
            .map(IRecipeViewer::favorites)
            .orElse(mutableListOf())

    /** @return Pair<Boolean: Stack, Boolean: ToInv> */
    @JvmStatic
    fun matchesKey(mouseKey: Int): Pair<Boolean, Boolean>? {
        return viewer
            .map { viewer: IRecipeViewer -> viewer.matchesKey(mouseKey) }
            .orElse(null)
    }

    @JvmStatic
    val isCheatMode: Boolean
        get() = viewer.map(IRecipeViewer::isCheatMode).orElse(false)

    @JvmStatic
    fun addFavorite(stack: GenericStack) {
        viewer.ifPresent { viewer: IRecipeViewer -> viewer.addFavorite(stack) }
    }

    @JvmStatic
    fun setSearchText(text: String) {
        viewer.ifPresent { viewer: IRecipeViewer -> viewer.setSearch(text) }
    }
}
